src/Function.cpp

```diff
@@ -71,12 +71,70 @@ using namespace std;
 
 ///////////////////////////////////////////////////////////////////////////////
 
+static vector<FunctionAttribute*> attributes;
 static vector<Function*> FuncList;		// List of all functions in the program
 static vector<FactMgr*>  FMList;        // list of fact managers for each function
 static long cur_func_idx;				// Index into FuncList that we are currently working on
 static bool param_first=true;			// Flag to track output of commas
 static int builtin_functions_cnt;
 
+FunctionAttribute::FunctionAttribute(string name, int prob)
+	:  attribute(name), attribute_probability(prob)
+{
+	//Nothing else to do
+}
+
+void
+FunctionAttribute::OutputAttribute(std::ostream &out, string option)
+{
+	out << " __attribute__((" << attribute << option << "))";
+}
+
+BooleanFunctionAttribute::BooleanFunctionAttribute(string name, int prob)
+	: FunctionAttribute(name, prob)
+{
+	//Nothing else to do
+}
+
+void
+BooleanFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(attribute_probability))
+		OutputAttribute(out, "");
+}
+
+MultiValuedFunctionAttribute::MultiValuedFunctionAttribute(string name, int prob, vector<string> arguments)
+	: FunctionAttribute(name, prob), attribute_values(arguments)
+{
+	//Nothing else to do
+}
+
+void
+MultiValuedFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(attribute_probability))
+		OutputAttribute(out, "(\"" +attribute_values[rnd_upto(attribute_values.size())] + "\")");
+}
+
+void
+Function::GenerateAttributes()
+{
+	if(CGOptions::func_attr_flag()){
+		vector<string> common_func_attributes = {"artificial", "flatten", "no_reorder", "hot", "cold", "noipa", "nothrow", "deprecated", "no_icf", "no_profile_instrument_function", "no_instrument_function", "no_sanitize_address", "no_sanitize_thread", "no_sanitize_undefined", "no_split_stack", "noinline", "noplt", "stack_protect"};
```

Comment:
This line is too long. Please split it. Take a look if the project has a line width limit?